### PR TITLE
Add coverage-touching and self-referential assertion heuristics

### PR DESCRIPTION
**Pull request by Evangelink**
## Summary

Adds two new named heuristics that close detection gaps in our test quality skills:

### Coverage touching (test-anti-patterns — Critical)
Tests that methodically call every public method on a type — often in alphabetical or declaration order — without asserting meaningful outcomes. The intent is to inflate code-coverage metrics rather than verify behavior. Distinct from "no assertions" because the pattern is *systematic* surface-area coverage with no real verification.

### Self-referential assertion (test-anti-patterns — Critical + assertion-quality metric)
Asserts that the output of an operation equals its input when the operation is expected to be identity or no-op (e.g. `Assert.AreEqual(input, Parse(input.ToString()))` or `Assert.AreEqual(dto.Name, dto.Name)`). The test is tautological — it can only fail if the round-trip is broken, but it never verifies that a *transformation* actually happened.

## Changes

**Skills:**
- `test-anti-patterns/SKILL.md` — Two new Critical-severity entries in the anti-pattern catalog
- `assertion-quality/SKILL.md` — New "Tests with self-referential assertions" metric + calibration rule

**Evaluations:**
- `tests/dotnet-test/test-anti-patterns/eval.yaml` — Two new scenarios:
  - Coverage-touching `ProductCatalog` facade (10 methods, systematic `_Works` naming, no/trivial assertions)
  - Self-referential `SerializerTests` (mix of valid round-trips and tautological identity tests)
- `tests/dotnet-test/assertion-quality/eval.yaml` — One new scenario:
  - Self-referential `ConfigTests` fixture with identity/round-trip patterns
- New fixture: `tests/dotnet-test/assertion-quality/fixtures/self-referential/ConfigService.Tests/ConfigTests.cs`

### CODE REVIEW

**Review by JanKrivanek — APPROVED** [review]

File: tests/dotnet-test/test-anti-patterns/fixtures/well-written/Payments.Tests/PaymentProcessorTests.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    [TestMethod]
+    public void ProcessPayment_NegativeAmount_ThrowsArgumentException()
+    {
+        var processor = new PaymentProcessor(new FakeGateway());
+
+        Assert.ThrowsException<ArgumentOutOfRangeException>(
+            () => processor.Process(new Payment("order-3", -10m)));
+    }
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply tests/dotnet-test/test-anti-patterns/fixtures/well-written/Payments.Tests/PaymentProcessorTests.cs:35]: Addressed in f298264a2 — renamed test methods to `ProcessPayment_ZeroAmount_ThrowsArgumentOutOfRangeException` / `ProcessPayment_NegativeAmount_ThrowsArgumentOutOfRangeException` to match the `ArgumentOutOfRangeException` assertion.

File: tests/dotnet-test/test-anti-patterns/fixtures/self-referential/Serialization.Tests/Serialization.Tests.csproj [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>net10.0</TargetFramework>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="MSTest" Version="3.*" />
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply tests/dotnet-test/test-anti-patterns/fixtures/self-referential/Serialization.Tests/Serialization.Tests.csproj]: Addressed in d93f0d46f — pinned MSTest to `3.8.2` and added `ImplicitUsings`/`Nullable`/`IsPackable` to the fixture .csproj.

File: tests/dotnet-test/test-anti-patterns/fixtures/well-written/Payments.Tests/Payments.Tests.csproj [hunk 3]
```diff
@@ -0,0 +1,8 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>net10.0</TargetFramework>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="MSTest" Version="3.*" />
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply tests/dotnet-test/test-anti-patterns/fixtures/well-written/Payments.Tests/Payments.Tests.csproj]: Addressed in d93f0d46f — pinned MSTest to `3.8.2` and added `ImplicitUsings`/`Nullable`/`IsPackable` to the fixture .csproj.

File: tests/dotnet-test/test-anti-patterns/eval.yaml [hunk 4]
```diff
@@ -126,6 +126,18 @@ scenarios:
...
       }
       ```
+    setup:
+      files:
+        - path: "UserRepo.Tests/UserRepo.Tests.csproj"
+          content: |
+            <Project Sdk="Microsoft.NET.Sdk">
+              <PropertyGroup><TargetFramework>net10.0</TargetFramework></PropertyGroup>
+              <ItemGroup><PackageReference Include="MSTest" Version="3.*" /></ItemGroup>
+            </Project>
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply tests/dotnet-test/test-anti-patterns/eval.yaml]: Addressed in d93f0d46f (inline csproj → `source:` reference) and superseded in 367cdaf59 (the inline-code prompts were removed entirely; all eval scenarios now load fixtures from disk via `setup.files`/`environment.files`).

File: tests/dotnet-test/assertion-quality/eval.vally.yaml [hunk 5]
```diff
@@ -90,6 +90,50 @@ stimuli:
...
+      files:
+        - src: fixtures/self-referential/ConfigService.Tests/ConfigTests.cs
+          dest: ConfigService.Tests/ConfigTests.cs
+        - content: |
+            <Project Sdk="Microsoft.NET.Sdk">
+              <PropertyGroup>
+                <TargetFramework>net10.0</TargetFramework>
+              </PropertyGroup>
+              <ItemGroup>
+                <PackageReference Include="MSTest" Version="3.*" />
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply tests/dotnet-test/assertion-quality/eval.vally.yaml]: Addressed in d93f0d46f (inline csproj → `source:` reference) and superseded in 367cdaf59 (the inline-code prompts were removed entirely; all eval scenarios now load fixtures from disk via `setup.files`/`environment.files`).

File: tests/dotnet-test/test-anti-patterns/fixtures/mixed-severity/UserRepo.Tests/UserRepo.Tests.csproj [hunk 6]
```diff
@@ -0,0 +1,8 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>net10.0</TargetFramework>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="MSTest" Version="3.*" />
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply tests/dotnet-test/test-anti-patterns/fixtures/mixed-severity/UserRepo.Tests/UserRepo.Tests.csproj]: Addressed in d93f0d46f — pinned MSTest to `3.8.2` and added `ImplicitUsings`/`Nullable`/`IsPackable` to the fixture .csproj.

File: tests/dotnet-test/test-anti-patterns/eval.yaml [hunk 7]
```diff
@@ -245,6 +253,16 @@ scenarios:
           }
       }
       ```
+    setup:
+      files:
+        - path: "FlakyCoupled.Tests/FlakyCoupled.Tests.csproj"
+          content: |
+            <Project Sdk="Microsoft.NET.Sdk">
+              <PropertyGroup><TargetFramework>net10.0</TargetFramework></PropertyGroup>
+              <ItemGroup><PackageReference Include="MSTest" Version="3.*" /></ItemGroup>
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply tests/dotnet-test/test-anti-patterns/eval.yaml]: Addressed in d93f0d46f (inline csproj → `source:` reference) and superseded in 367cdaf59 (the inline-code prompts were removed entirely; all eval scenarios now load fixtures from disk via `setup.files`/`environment.files`).

File: tests/dotnet-test/assertion-quality/eval.yaml [hunk 8]
```diff
@@ -87,7 +87,45 @@ scenarios:
...
+    setup:
+      files:
+        - path: "ConfigService.Tests/ConfigService.Tests.csproj"
+          content: |
+            <Project Sdk="Microsoft.NET.Sdk">
+              <PropertyGroup>
+                <TargetFramework>net10.0</TargetFramework>
+              </PropertyGroup>
+              <ItemGroup>
+                <PackageReference Include="MSTest" Version="3.*" />
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply tests/dotnet-test/assertion-quality/eval.yaml]: Addressed in d93f0d46f — pinned MSTest to `3.8.2` and added `ImplicitUsings`/`Nullable`/`IsPackable` to the fixture .csproj.

[the remark this answers is not in the document]
Comment (Evangelink) [reply tests/dotnet-test/assertion-quality/eval.yaml]: Addressed in d93f0d46f (inline csproj → `source:` reference) and superseded in 367cdaf59 (the inline-code prompts were removed entirely; all eval scenarios now load fixtures from disk via `setup.files`/`environment.files`).

File: tests/dotnet-test/test-anti-patterns/fixtures/coverage-touching/CatalogService.Tests/CatalogService.Tests.csproj [hunk 9]
```diff
@@ -0,0 +1,8 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>net10.0</TargetFramework>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="MSTest" Version="3.*" />
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply tests/dotnet-test/test-anti-patterns/fixtures/coverage-touching/CatalogService.Tests/CatalogService.Tests.csproj]: Addressed in d93f0d46f — pinned MSTest to `3.8.2` and added `ImplicitUsings`/`Nullable`/`IsPackable` to the fixture .csproj.

**Comment by Evangelink** [comment]
Pushed `367cdaf59` to address all remaining issues from PR #635 (eval scoreboard run [25677669933](https://github.com/dotnet/skills/actions/runs/25677669933) + skill-check job [25810696119](https://github.com/dotnet/skills/actions/runs/25810696119)).

## What changed

### 1. test-anti-patterns activation (was NOT ACTIVATED in **both** plugin and isolated mode for **all 6 scenarios**)
The verdict was unambiguous: model answered every scenario from the inline ``csharp`` block in the prompt and never needed the skill. Even isolated mode failed to activate.

- `tests/dotnet-test/test-anti-patterns/eval.yaml` and `eval.vally.yaml` — dropped the giant inline code blocks; each prompt now names the workspace files explicitly (e.g. `Audit ``UserRepo.Tests/UserRepositoryTests.cs`` for .NET test anti-patterns ...`) and asks for the skill's signature deliverable (severity-ranked list with concrete code-level fixes).
- All 6 scenarios consistently get `reject_tools: ["bash", "edit", "create"]` (read-only review) and `timeout: 180`.
- `eval.vally.yaml` also got missing `environment.files` for each scenario (was prompt-only before — no fixture on disk).

### 2. assertion-quality `Flag assertion-free tests` plugin-mode activation
- Rewrite the prompt to name the file (`SmokeTests/ApiEndpointSmokeTests.cs`) and explicitly request `assertion quality and depth`, `assertion-free tests`, and `trivial assertions like Assert.IsNotNull` — direct trigger phrases for assertion-quality, disambiguating from sibling skills like test-anti-patterns.
- Strengthen `assertion-quality` SKILL.md description with explicit USE FOR keywords for `assertion-free tests`, `self-referential`, and `tautological`.

### 3. test-anti-patterns description size (skill-check failure)
The previous activation fix grew the description to **1,301 chars** (over the 1,024 per-skill limit) and pushed the dotnet-test plugin aggregate to **15,312** (over the 15,000 limit). Trimmed to **779 chars**; aggregate now at **14,983**.

### 4. Vally job-level timeout (60m → 180m)
Already raised on `main` in commit 0f31228b5 (#641). Inherited via the merge of `upstream/main`.

### 5. Eval scenario timeouts
- `assertion-quality / Decline request to write new tests from scratch` was already bumped to 300s in this PR (commit bbcd96fb2). The 120s ⏰ in the scoreboard is from a stale eval.yaml at run time.
- All test-anti-patterns scenarios bumped from 120s to 180s for headroom.

### 6. Open Copilot review comments
All 10 unresolved review threads addressed and replied to (the underlying changes were already in commits f298264a2 + d93f0d46f; commit 367cdaf59 also makes the inline-csproj concerns moot by removing all inline-code prompts).

## Verification

- `dotnet run --project eng/skill-validator/src/SkillValidator.csproj -- check --plugin ./plugins/dotnet-test` ✅ (22 skills, 11 agents, 1 plugin)
- Aggregate dotnet-test plugin description: 14,983 / 15,000
- `test-anti-patterns` description: 779 / 1,024
- `assertion-quality` description: 691 / 1,024
- `markdownlint-cli2` clean on touched SKILL.md files
- All 4 eval YAMLs parse with PyYAML

File: tests/dotnet-test/test-anti-patterns/fixtures/self-referential/Serialization.Tests/SerializerTests.cs [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    [TestMethod]
+    public void UpperCase_ToUpper_Works()
+    {
+        var input = "HELLO";
+        var result = input.ToUpper();
+        Assert.AreEqual(input, result);
+    }
+}
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply tests/dotnet-test/test-anti-patterns/fixtures/self-referential/Serialization.Tests/SerializerTests.cs:78]: Addressed in 967f0ad97 — added `fixtures/self-referential/Serialization/Serialization.cs` with minimal stubs for all 8 referenced types (`Serializer`, `UserDto`, `Money`, `Config`, `NameNormalizer`, `Mapper`, `UserEntity`, `Validator`). Updated the `Detect self-referential...` scenario in eval.yaml + eval.vally.yaml to copy the new file into the workspace.

File: tests/dotnet-test/test-anti-patterns/fixtures/coverage-touching/CatalogService.Tests/ProductCatalogTests.cs [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+[TestClass]
+public class ProductCatalogTests
+{
+    private ProductCatalog _catalog = null!;
+
+    [TestInitialize]
+    public void Setup()
+    {
+        _catalog = new ProductCatalog(new InMemoryStore());
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply tests/dotnet-test/test-anti-patterns/fixtures/coverage-touching/CatalogService.Tests/ProductCatalogTests.cs:13]: Addressed in 967f0ad97 — added a minimal `InMemoryStore : IProductStore` implementation in `ProductCatalog.cs` so audit agents see a complete, compilable fixture and stay focused on the coverage-touching pattern.

File: tests/dotnet-test/assertion-quality/eval.yaml [hunk 12]
```diff
@@ -87,7 +92,37 @@ scenarios:
...
+        pattern: "(GetTimeout|GetRetries|FormatConnectionString|ValidateKey)"
+      - type: "exit_success"
+    rubric:
+      - "Identified that multiple tests assert the output equals the input — a self-referential pattern that cannot catch transformation bugs"
+      - "Flagged GetTimeout_Default_ReturnsDefault and GetRetries_Default_ReturnsDefault as tautological — they assert that reading back a just-written value returns the same value without testing the default/fallback behavior the names suggest"
+      - "Flagged FormatConnectionString_RoundTrip as only proving round-trip identity without verifying the formatted string's actual structure or content"
+      - "Flagged ValidateKey_Valid_ReturnsSameKey as not exercising the validation logic — it only proves valid input passes through unchanged"
+      - "Noted that Serialize_Deserialize round-trip is potentially valid but should be accompanied by tests with edge-case inputs (special characters, nulls, empty strings)"
+      - "Recommended concrete improvements: test actual default values, test invalid keys, test formatted string structure, test edge-case serialization inputs"
+    timeout: 300
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply tests/dotnet-test/assertion-quality/eval.yaml:129]: Addressed in 967f0ad97 — added `reject_tools: [bash, edit, create]` (and `constraints.reject_tools` in eval.vally.yaml) plus explicit `Do not edit the files` wording to the `Identify self-referential...` scenario, matching the other read-only audit scenarios.
